### PR TITLE
Allow to control sealing module visibility

### LINKED ISSUE #9: Allow to control visibility
While using `pub(crate) mod` allows to overcome nesting for impls, it also makes using `#[sealed]` pointless in bounds of a single crate: imagine usage in application crate, where the trait is declared in some module, not other crate.

It would be nice to control the visibility via attribute argument:
```rust
#[sealed]
pub trait T {}
// expands to
mod __seal_t {}

#[sealed(pub(crate))]
pub trait T {}
// expands to
pub(crate) mod __seal_t {}

// And even follow an arbitrary visiblity:
// https://doc.rust-lang.org/reference/visibility-and-privacy.html

#[sealed(pub(in crate::very::long::path))]
pub trait T {}
// expands to
pub(in crate::very::long::path) mod __seal_t {}
```

So, for the nesting case (which is quite rare, AFAIK), an user should specify the visibility explicitly, while for the usage in non-library crate, the sealing will work well out-of-the-box.

### PR DESCRIPTION
**Pull request by tyranron**
Fixes #9 

Bumps up crate to `0.3.0` version as using private visibility by default is a breaking change.

- [x] Tests added
- [x] Documentation and README are updated
- [x] Examples and demo are updated

### CODE REVIEW

**Comment by tyranron** [comment]
@jmg-duarte please, review.

File: src/lib.rs [hunk 1]
```diff
@@ -185,3 +213,78 @@ fn parse_sealed_impl(item_impl: &syn::ItemImpl) -> syn::Result<TokenStream2> {
...
+                    syn::Ident::parse_any(input)?;
+                    out.erased = true;
+                }
+
+                "pub" => {
+                    out.visibility = input.parse()?;
+                    if matches!(out.visibility, syn::Visibility::Public(_)) {
+                        return Err(syn::Error::new(
+                            out.visibility.span(),
+                            "seal with `pub` visibility is nonsense",
```
Comment (jmg-duarte) [inline src/lib.rs:263]: Provide a reason as to why it is "nonsense" instead

**Review by jmg-duarte — CHANGES_REQUESTED** [review]
IMO all error messages should be capitalized, if you disagree, I am open to know more as to why

**Comment by tyranron** [comment]
@jmg-duarte 

> IMO all error messages should be capitalized, if you disagree, I am open to know more as to why

Just following the style of standard compile-time error messages (like in `tests/fail/*.stderr` files). I tend to capitilze error messages too, but rather runtime's ones.

**Comment by tyranron** [comment]
@jmg-duarte CI fails due to nightly Rust printing slightly changed diagnostic messages than stable.

Some approaches:
1. We can opt-out codegen failure tests for nightly.
2. We can copy-paste the suite for nightly, correct the `.stderr` files and run different suits on nightly and stable.

What should we prefer?

**Comment by jmg-duarte** [comment]
> @jmg-duarte
> 
> > IMO all error messages should be capitalized, if you disagree, I am open to know more as to why
> 
> Just following the style of standard compile-time error messages (like in `tests/fail/*.stderr` files). I tend to capitilze error messages too, but rather runtime's ones.

Great point, I hadn't noticed before

**Comment by jmg-duarte** [comment]
> @jmg-duarte CI fails due to nightly Rust printing slightly changed diagnostic messages than stable.
> 
> Some approaches:
> 
>     1. We can opt-out codegen failure tests for nightly.
> 
>     2. We can copy-paste the suite for nightly, correct the `.stderr` files and run different suits on nightly and stable.
> 
> 
> What should we prefer?

I've opted to only do UI tests on stable (https://github.com/jmg-duarte/sealed-rs/issues/11), I'll be working on this in a bit

File: src/lib.rs (hunk 1 above)
Comment (tyranron) [reply src/lib.rs:263]: @jmg-duarte is something like 
```
`pub` visibility cannot seal anything
```
will be acceptable enough?

Comment (jmg-duarte) [reply src/lib.rs:263]: What about
```
`pub` visibility is not strict enough. Consider refining the visibility level to something stricter (e.g. `pub(crate)`) or if you actually need sealing.
```
Possibly, we could add a link to the API Guidelines book?

Comment (tyranron) [reply src/lib.rs:263]: @jmg-duarte hmmm... something like this? 

```
error: `pub` visibility allows to use the seal outside its crate.
Consider tightening the visibility level (e.g. `pub(crate)`) if you actually need sealing.
Read more here: https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed
```

**Comment by jmg-duarte** [comment]
@tyranron the new CI *should* be able to cope with your changes

File: src/lib.rs (hunk 1 above)
Comment (jmg-duarte) [reply src/lib.rs:263]: Yes! But the link may be too much, now that I see it rendered

What are your thoughts on it?

Comment (tyranron) [reply src/lib.rs:263]: @jmg-duarte I agree to go without the link.

Comment (jmg-duarte) [reply src/lib.rs:263]: Then I think you just need to push the changes and hopefully the CI won't raise problems with the UI tests

**Comment by jmg-duarte** [comment]
Everything checks out, however I noticed one thing that is missing. 

What is the default visibility when the argument is not passed in? Could you document that?

**Comment by tyranron** [comment]
@jmg-duarte I've thought this is quite obvious from the expansion examples. But mentioning defaults is not a big deal, so here we go.

**Comment by jmg-duarte** [comment]
> @jmg-duarte I've thought this is quite obvious from the expansion examples. But mentioning defaults is not a big deal, so here we go.

*Explicit > Implicit*

Regardless, great work! Thank you for your contribution!

**Pull request merged**